### DISCUSSION TITLE
Unit test case writing issue for covering Instanceof code

### DISCUSSION THREAD

**Discussion by suragar** [SalesforceCommerceCloud/sfcc-ci#402] [Team Posts]
Hi

We are trying to write unit test case in SFRA for a piece of code where a variable is tested as instanceof another demandware object. Please let me know how to resolve the same.

Code snippet:

if (lineItem instanceof dw.order.ProductLineItem && !lineItem.optionProductLineItem && lineItem.custom.isThisBillable) {
                lineStatus = lineItem.custom.sapLineStatus;
                if (lineStatus && lineStatus == ORDR_STATUS_SHIPPED) {
                    shippedQuantity += 1;
                } else if (lineStatus && lineStatus == ORDR_STATUS_RETURNED) {
                    returnedQuantity += 1;
                } else if (lineStatus && lineStatus == ORDR_STATUS_CANCELLED) {
                    cancelledQuantity += 1;
                }
            }

the if condition never matches.

**Comment by zlatinz** [comment]
Instead of `lineItem instanceof dw.order.ProductLineItem`, why not you check for `lineItem.catalogProduct`?
And in your mock classes you make sure you have `catalogProduct` property.
Unit tests are not executed against the actual API classes (hence checks like `instanceof dw.package.ScriptAPIClassName` will always fail), they are using mocks to simulate the data.

**Comment by suragar** [comment]
We have the code written and its working absolutely fine, we need to write the unit test case, so please suggest how to resolve the dependency for ProductLineItem so our unit test case can cover these lines.

Thanks in advance.

**Comment by zlatinz** [comment]
Yes, code is working fine, but you cannot mock `dw.order.ProductLineItem` from the unit test framework, so that it is recognized by your code, when it is run in the context of unit testing.
You need to change the code to either do check in a different way, or check for instanceof on a require-d class.

E.g.
``` 
var ProductLineItem = require('dw/order/ProductLineItem');

// ...
if (lineItem instanceof ProductLineItem && !lineItem.optionProductLineItem && lineItem.custom.isThisBillable) {
// ...
```

Then in the unit test you need to provide a mock for require('dw/order/ProductLineItem') (and also ensure that lineItem is instance of the same mock class)

**Comment by suragar** [comment]
This is another way we have implemented, but still our unit test case is not covering this piece of if condition. We think that the dependency is not getting resolved as per expectation for ProductLineItem to fulfill this condition. But again we dont know how to do the same. Any suggestions?

**Comment by zlatinz** [comment]
How is your unit test code looking?

**Comment by pravngaur** [comment]
Could you please provide the error logs - if your script is failing(which
should be if it is not able to resolve the dependencies), that would help
in understanding the issue.

On Thu, May 30, 2019 at 6:09 PM Surbhi Agarwal <notifications@github.com>
wrote:

> This is another way we have implemented, but still our unit test case is
> not covering this piece of if condition. We think that the dependency is
> not getting resolved as per expectation for ProductLineItem to fulfill this
> condition. But again we dont know how to do the same. Any suggestions?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/SalesforceCommerceCloud/teams/community/discussions/48/comments/4?email_source=notifications&email_token=AKECNZPBMAGFBXYBP4R7GPTPX7DG5A5CNFSM4HRCJCTKYY3PNVWWK3TUL52HS4DFWNCGS43DOVZXG2LPNZIG643UKJSXA3DZVJRW63LNMVXHIX3JMTHAAAPJIE>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKECNZM7HGLRXTVGDWJRPL3PX7DG5ANCNFSM4HRCJCTA>
> .
>


-- 
Thanks and Regards,
Praveen Gaur
+91-9953335721

**Comment by suragar** [comment]
Hi

There are no error logs. The script is not failing, the problem is that this piece of code is not getting covered as part of coverage.

Thanks
Surbhi Agarwal

Sent from mobile device, all errors self inflicted

________________________________

**Comment by pravngaur** [comment]
Hi, which if condition is not matching 1 or 2nd?

Also, is the lineItem is actually a line item &
lineItem.custom.isThisBillable true?

Are you getting any error in logs(like null pointer) - which will help you
in identifying the root cause.

Or, if the test should be passed as long as it is a line item, then you can
also use the alternative suggested by Zaltin.
